### PR TITLE
RegexFullMatch operator

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Motivation and Context
Closes https://github.com/microsoft/onnxruntime/issues/17594.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (xadupre) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:30]: Should we parallelize?

Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/nn/regex_full_match.cc:30]: We could do, it looks like the RE2 object is thread safe as well. Are there any pointers on how to use the ORT threadpool?

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:21]: This input is not optional

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:25]: If this fails, then the index is wrong, so it must be `ORT_ENFORCE` or assert. However, in practice we do not do that.

File: onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc
Comment (yuslepukhin) [inline onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc:15]: I would like to see some UTF-8 multibyte characters tests.
Please, add tests for French, German, Russian and Chinese.

For French strings include une cédille like in Besançon.
German umlaut like in Mit freundlichen Grüßen.
Any Cyrillic for Russian is fine such a as 'Понедельник'.
Any Kanji is fine or you may use any Japanese characters from Hiragana or Katakana.

**Review by yuslepukhin — CHANGES_REQUESTED** [review]

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:31]:
```suggestion
  for (auto i = 0, limit = input_tensor->Shape().Size(); i < limit; ++i) {
    output_data[i] = RE2::FullMatch(input_data[i], re);
```

Comment (yuslepukhin) [reply onnxruntime/core/providers/cpu/nn/regex_full_match.cc:31]: Refrain from using int64_t as operator[] index.

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:19]: can we initialize a RE2 data member in the ctor and reuse it instead of creating one every call to Compute()?

File: onnxruntime/core/providers/cpu/nn/regex_full_match.cc
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:1]:
```suggestion
#include "core/providers/cpu/nn/regex_full_match.h"
```

Comment (edgchen1) [inline onnxruntime/core/providers/cpu/nn/regex_full_match.cc:1]: general: add copyright header

File: onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc
Comment (adityagoel4512) [reply onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc:14]: > Also, it is interesting to see what happens if you try to parse invalid UTF-8 data, and see if we could emit an understandable error message that would make it easy to deal with support cases.

So we raise an error stating that an invalid regex pattern has been provided. In addition the following is logged by re2:

`/<path to build>/_deps/re2-src/re2/re2.cc:205: Error parsing '��ABÀ���': invalid UTF-8`

File: onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc
Comment (yuslepukhin) [inline onnxruntime/test/providers/cpu/nn/regex_full_match_test.cc:25]: The actual character to test is `ç`

**Comment by thiagocrepaldi** [comment]
@adityagoel4512 any update on this? ORT 1.17 release is coming soon :)

**Comment by adityagoel4512** [comment]
> @adityagoel4512 any update on this? ORT 1.17 release is coming soon :)

Waiting on a review round/CI pipelines to be run by a maintainer.

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by yuslepukhin** [comment]
Changes to docs required:
[REgFullMatchDocs.txt](https://github.com/microsoft/onnxruntime/files/13893412/REgFullMatchDocs.txt)

**Comment by yuslepukhin** [comment]
WASM build tests fail, will try to find out what is going on:

> 2024-01-10T19:49:29.2164985Z 1: [32m10 01 2024 19:49:29.214:INFO [karma-server]: [39mKarma v6.4.1 server started at http://localhost:9876/
2024-01-10T19:49:29.2165843Z 1: [32m10 01 2024 19:49:29.215:INFO [launcher]: [39mLaunching browsers ChromeHeadlessTest with concurrency unlimited
2024-01-10T19:49:29.2184593Z 1: [32m10 01 2024 19:49:29.218:INFO [launcher]: [39mStarting browser ChromeHeadless
2024-01-10T19:49:33.6317399Z 1: [32m10 01 2024 19:49:33.631:INFO [Chrome Headless 120.0.6099.199 (Linux x86_64)]: [39mConnected on socket wzGXfitT0_D6Qz5sAAAB with id 54354320
2024-01-10T19:49:35.0661766Z 1: Chrome Headless 120.0.6099.199 (Linux x86_64): Executed 0 of 1 SUCCESS (0 secs / 0 secs)
2024-01-10T19:49:37.4509002Z 1: [1A[2KERROR: 'worker sent an error! http://localhost:9876/base/onnxruntime_test_all.worker.js:172: Uncaught RuntimeError: memory access out of bounds'
2024-01-10T19:49:37.4509827Z 1: Chrome Headless 120.0.6099.199 (Linux x86_64): Executed 0 of 1 SUCCESS (0 secs / 0 secs)
2024-01-10T19:49:37.4510388Z 1: [1A[2KChrome Headless 120.0.6099.199 (Linux x86_64) ERROR
2024-01-10T19:49:37.4510675Z 1:   Uncaught [object ErrorEvent]
2024-01-10T19:49:37.4510958Z 1:   at onnxruntime_test_all.js:6674:4
2024-01-10T19:49:37.4511290Z 1: Chrome Headless 120.0.6099.199 (Linux x86_64): Executed 0 of 1 ERROR (0 secs / 0 secs)

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Pull request merged**
